prow/sidecar/censor.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sidecar
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/semaphore"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"k8s.io/test-infra/prow/pod-utils/wrapper"
+	"k8s.io/test-infra/prow/secretutil"
+)
+
+// defaultFrameSize is the default frame size, 10MiB. We could have a smaller footprint
+// theoretically as long as the secrets we censor are small, but we risk not being performant
+// as the I/O cost would skyrocket and would not scale well.
+const defaultFrameSize = 10 * 1024 * 1024
+
+func (o Options) censor() error {
+	var concurrency int64
+	if o.CensoringConcurrency == nil {
+		concurrency = int64(runtime.GOMAXPROCS(0))
+	} else {
+		concurrency = *o.CensoringConcurrency
+	}
+	sem := semaphore.NewWeighted(concurrency)
+	wg := &sync.WaitGroup{}
+	errors := make(chan error)
+	var errs []error
+	go func() {
+		for err := range errors {
+			errs = append(errs, err)
+		}
+	}()
+
+	secrets, err := loadSecrets(o.Secrets)
+	if err != nil {
+		return fmt.Errorf("could not load secrets: %w", err)
+	}
+	censorer := secretutil.NewCensorer()
+	censorer.RefreshBytes(secrets...)
+
+	// We want to keep as little of the file in memory as possible in order for
+	// this operation to be reasonably performant in space. However, to guarantee
+	// that we censor every instance of every secret, our frame buffer must be at
+	// least two times larger than the largest secret we are about to censor.
+	frameSize := defaultFrameSize
+	if largest := censorer.LargestSecret(); 2*largest > frameSize {
```

Comment:
this logic is effectively untested, maybe add a private variadic option to override the default?